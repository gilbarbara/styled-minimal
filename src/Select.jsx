import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { sizeType } from './utils/propTypes';
import { FormStyles } from './utils/system';

export const StyledSelect = styled.select`
  ${FormStyles.select};
  ${FormStyles.pseudo};
`;

const Select = ({ children, ...props }) => (
  <StyledSelect {...props}>{children}</StyledSelect>
);

Select.propTypes = {
  children: PropTypes.node.isRequired,
  disabled: PropTypes.bool,
  multiple: PropTypes.bool,
  required: PropTypes.bool,
  size: PropTypes.number,
  sizing: sizeType,
};

Select.defaultProps = {
  sizing: 'md',
};

export default Select;
