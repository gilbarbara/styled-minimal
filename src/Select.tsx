import * as React from 'react';
import styled, { css } from 'styled-components';

import { getTheme, px } from './utils/helpers';
import { baseStyles, pseudoStyles, sx, system } from './utils/system';

import { SelectProps } from './types';

export const StyledSelect = styled('select')<SelectProps>(
  props => {
    const { multiple, sizing = 'md', valid } = props;
    const {
      backgroundColor,
      color,
      borderColor,
      borderRadius,
      borderWidth,
      fontSize,
      height,
      lineHeight,
      padding,
      validation,
    } = getTheme(props, 'select');

    const currentBgColor = backgroundColor;
    let currentBorderColor = borderColor;

    if (valid) {
      currentBorderColor = validation.valid;
    } else if (valid === false) {
      currentBorderColor = validation.invalid;
    }

    const multipleCheckedColor = currentBorderColor;

    return css`
      ${baseStyles};
      background-color: ${currentBgColor};
      border: ${px(borderWidth)} solid ${currentBorderColor};
      border-radius: ${px(borderRadius)};
      color: ${color};
      display: block;
      ${!multiple ? `height: ${px(height[sizing])}` : ''};
      font-family: inherit;
      font-size: ${px(fontSize[sizing])};
      line-height: ${lineHeight};
      ${!multiple ? `padding: ${px(padding[sizing])};` : ''};
      white-space: nowrap;
      width: 100%;

      > option {
        background-color: ${currentBgColor};
        color: ${color};
        font-size: ${px(fontSize[sizing])};
        line-height: ${lineHeight};
        padding: ${px(padding[sizing])};
      }

      &[multiple] {
        background-color: ${currentBgColor};

        option:checked {
          background: ${`${multipleCheckedColor} linear-gradient(0deg, ${multipleCheckedColor} 0%, ${multipleCheckedColor} 100%)`};
          font-weight: bold;
        }
      }
    `;
  },
  pseudoStyles,
  sx,
  system,
);

export const Select = React.forwardRef<HTMLSelectElement, SelectProps>((props, ref) => (
  <StyledSelect ref={ref} {...props} />
));

Select.defaultProps = {
  sizing: 'md',
};
