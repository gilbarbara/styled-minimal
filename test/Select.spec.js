import React from 'react';
import { mount } from 'enzyme';

import Select from '../src/Select';

describe('Select', () => {
  const wrapper = mount(
    <Select>
      <option value="male">Male</option>
      <option value="female">Female</option>
      <option value="other">Other</option>
    </Select>
  );

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
