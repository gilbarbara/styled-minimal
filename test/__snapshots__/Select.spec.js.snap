// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select should render properly 1`] = `
.c0 {
  box-sizing: border-box;
  background-color: #fff;
  border: 1px solid #ccc;
  border-radius: 2px;
  color: #343434;
  display: block;
  height: 34px;
  font-family: inherit;
  font-size: 16px;
  line-height: 1;
  padding: 8px;
  white-space: nowrap;
  width: 100%;
}

.c0 > option {
  background-color: #fff;
  font-size: 16px;
  line-height: 1;
  padding: 8px;
}

.c0[multiple]:focus option:checked {
  background: black linear-gradient(0deg,black 0%,black 100%);
}

.c0::-webkit-input-placeholder {
  color: #b3b3b3;
}

.c0:-moz-placeholder {
  color: #b3b3b3;
}

.c0::-moz-placeholder {
  color: #b3b3b3;
}

.c0:-ms-input-placeholder {
  color: #b3b3b3;
}

.c0:-moz-read-only {
  background-color: #fafafa;
  color: #818181;
}

.c0:read-only {
  background-color: #fafafa;
  color: #818181;
}

.c0:disabled {
  background-color: #f2f2f2;
  color: #676767;
}

.c0:focus {
  outline-color: #000;
}

.c0:required:not(:valid) {
  border-color: #000;
}

.c0:required:not(:valid)::-webkit-input-placeholder {
  color: #000;
}

.c0:required:not(:valid):-moz-placeholder {
  color: #000;
}

.c0:required:not(:valid)::-moz-placeholder {
  color: #000;
}

.c0:required:not(:valid):-ms-input-placeholder {
  color: #000;
}

<Select
  as="select"
  sizing="md"
>
  <select
    className="c0"
  >
    <option
      value="male"
    >
      Male
    </option>
    <option
      value="female"
    >
      Female
    </option>
    <option
      value="other"
    >
      Other
    </option>
  </select>
</Select>
`;
